File: src/test/java/com/syfe/personalfinance/service/SavingsGoalServiceTest.java
package com.syfe.personalfinance.service;

import com.syfe.personalfinance.dto.SavingsGoalDto;
import com.syfe.personalfinance.dto.SavingsGoalDto.GoalProgressResponse;
import com.syfe.personalfinance.entity.SavingsGoal;
import com.syfe.personalfinance.entity.User;
import com.syfe.personalfinance.enums.CategoryType;
import com.syfe.personalfinance.exception.BadRequestException;
import com.syfe.personalfinance.mapper.SavingsGoalMapper;
import com.syfe.personalfinance.repository.SavingsGoalRepository;
import com.syfe.personalfinance.repository.TransactionRepository;
import com.syfe.personalfinance.service.impl.SavingsGoalServiceImpl;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class SavingsGoalServiceTest {

    @Mock
    private SavingsGoalRepository savingsGoalRepository;

    @Mock
    private TransactionRepository transactionRepository;

    @Mock
    private SavingsGoalMapper savingsGoalMapper;

    @Mock
    private UserService userService;

    private SavingsGoalService savingsGoalService;

    private User testUser;

    @BeforeEach
    void setUp() {
        savingsGoalService = new SavingsGoalServiceImpl(savingsGoalRepository, transactionRepository, savingsGoalMapper, userService);
        testUser = User.builder().id(100L).username("user@syfe.com").build();
    }

    @Test
    void createGoal_Success() {
        SavingsGoalDto.CreateGoalRequest request = SavingsGoalDto.CreateGoalRequest.builder()
                .goalName("Car fund")
                .targetAmount(new BigDecimal("10000.00"))
                .startDate(LocalDate.now())
                .targetDate(LocalDate.now().plusYears(1))
                .build();

        SavingsGoal goalEntity = SavingsGoal.builder()
                .goalName("Car fund")
                .targetAmount(new BigDecimal("10000.00"))
                .startDate(LocalDate.now())
                .targetDate(LocalDate.now().plusYears(1))
                .user(testUser)
                .build();

        GoalProgressResponse expectedResponse = GoalProgressResponse.builder()
                .id(1L)
                .goalName("Car fund")
                .targetAmount(new BigDecimal("10000.00"))
                .build();

        when(userService.getAuthenticatedUserEntity()).thenReturn(testUser);
        when(savingsGoalMapper.toEntity(request)).thenReturn(goalEntity);
        when(savingsGoalRepository.save(any(SavingsGoal.class))).thenReturn(goalEntity);
        when(savingsGoalMapper.toProgressResponse(any(SavingsGoal.class), any(), any(), any())).thenReturn(expectedResponse);

        GoalProgressResponse response = savingsGoalService.createGoal(request);

        assertNotNull(response);
        assertEquals("Car fund", response.getGoalName());
        verify(savingsGoalRepository, times(1)).save(any(SavingsGoal.class));
    }

    @Test
    void createGoal_TargetBeforeStart_ThrowsException() {
        SavingsGoalDto.CreateGoalRequest request = SavingsGoalDto.CreateGoalRequest.builder()
                .goalName("Car fund")
                .targetAmount(new BigDecimal("10000.00"))
                .startDate(LocalDate.now())
                .targetDate(LocalDate.now().minusDays(5))
                .build();

        when(userService.getAuthenticatedUserEntity()).thenReturn(testUser);

        assertThrows(BadRequestException.class, () -> savingsGoalService.createGoal(request));
        verify(savingsGoalRepository, never()).save(any(SavingsGoal.class));
    }

    @Test
    void getGoalProgress_SuccessCalculations() {
        LocalDate startDate = LocalDate.now().minusMonths(3);
        SavingsGoal goal = SavingsGoal.builder()
                .id(1L)
                .goalName("Car fund")
                .targetAmount(new BigDecimal("10000.00"))
                .startDate(startDate)
                .targetDate(LocalDate.now().plusYears(1))
                .user(testUser)
                .build();

        GoalProgressResponse expectedProgress = GoalProgressResponse.builder()
                .id(1L)
                .goalName("Car fund")
                .targetAmount(new BigDecimal("10000.00"))
                .currentProgress(new BigDecimal("3000.00"))
                .progressPercentage(new BigDecimal("30.00"))
                .remainingAmount(new BigDecimal("7000.00"))
                .build();

        when(userService.getAuthenticatedUserEntity()).thenReturn(testUser);
        when(savingsGoalRepository.findByIdAndUserId(1L, 100L)).thenReturn(Optional.of(goal));

        when(transactionRepository.sumAmountByUserIdAndCategoryTypeSinceDate(100L, CategoryType.INCOME, startDate))
                .thenReturn(new BigDecimal("5000.00"));
        when(transactionRepository.sumAmountByUserIdAndCategoryTypeSinceDate(100L, CategoryType.EXPENSE, startDate))
                .thenReturn(new BigDecimal("2000.00"));

        when(savingsGoalMapper.toProgressResponse(eq(goal), any(BigDecimal.class), any(BigDecimal.class), any(BigDecimal.class)))
                .thenReturn(expectedProgress);

        GoalProgressResponse progress = savingsGoalService.getGoalProgress(1L);

        assertNotNull(progress);
        assertEquals(new BigDecimal("3000.00"), progress.getCurrentProgress());
        assertEquals(new BigDecimal("30.00"), progress.getProgressPercentage());
        assertEquals(new BigDecimal("7000.00"), progress.getRemainingAmount());
    }
}

File: src/test/java/com/syfe/personalfinance/service/SavingsGoalServiceTest.kt
package com.syfe.personalfinance.service

import com.syfe.personalfinance.dto.SavingsGoalDto
import com.syfe.personalfinance.dto.SavingsGoalDto.GoalProgressResponse
import com.syfe.personalfinance.entity.SavingsGoal
import com.syfe.personalfinance.entity.User
import com.syfe.personalfinance.enums.CategoryType
import com.syfe.personalfinance.exception.BadRequestException
import com.syfe.personalfinance.mapper.SavingsGoalMapper
import com.syfe.personalfinance.repository.SavingsGoalRepository
import com.syfe.personalfinance.repository.TransactionRepository
import com.syfe.personalfinance.service.impl.SavingsGoalServiceImpl
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.any
import org.mockito.kotlin.eq
import org.mockito.kotlin.never
import org.mockito.kotlin.times
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import java.math.BigDecimal
import java.time.LocalDate
import java.util.Optional

@ExtendWith(MockitoExtension::class)
class SavingsGoalServiceTest {

    @Mock
    private lateinit var savingsGoalRepository: SavingsGoalRepository

    @Mock
    private lateinit var transactionRepository: TransactionRepository

    @Mock
    private lateinit var savingsGoalMapper: SavingsGoalMapper

    @Mock
    private lateinit var userService: UserService

    private lateinit var savingsGoalService: SavingsGoalService

    private lateinit var testUser: User

    @BeforeEach
    fun setUp() {
        savingsGoalService = SavingsGoalServiceImpl(savingsGoalRepository, transactionRepository, savingsGoalMapper, userService)
        testUser = User(id = 100L, username = "user@syfe.com")
    }

    @Test
    fun createGoal_Success() {
        val request = SavingsGoalDto.CreateGoalRequest(
            goalName = "Car fund",
            targetAmount = BigDecimal("10000.00"),
            startDate = LocalDate.now(),
            targetDate = LocalDate.now().plusYears(1)
        )

        val goalEntity = SavingsGoal(
            goalName = "Car fund",
            targetAmount = BigDecimal("10000.00"),
            startDate = LocalDate.now(),
            targetDate = LocalDate.now().plusYears(1),
            user = testUser
        )

        val expectedResponse = GoalProgressResponse(
            id = 1L,
            goalName = "Car fund",
            targetAmount = BigDecimal("10000.00")
        )

        whenever(userService.getAuthenticatedUserEntity()).thenReturn(testUser)
        whenever(savingsGoalMapper.toEntity(request)).thenReturn(goalEntity)
        whenever(savingsGoalRepository.save(any<SavingsGoal>())).thenReturn(goalEntity)
        whenever(savingsGoalMapper.toProgressResponse(any<SavingsGoal>(), any(), any(), any())).thenReturn(expectedResponse)

        val response = savingsGoalService.createGoal(request)

        assertNotNull(response)
        assertEquals("Car fund", response.goalName)
        verify(savingsGoalRepository, times(1)).save(any<SavingsGoal>())
    }

    @Test
    fun createGoal_TargetBeforeStart_ThrowsException() {
        val request = SavingsGoalDto.CreateGoalRequest(
            goalName = "Car fund",
            targetAmount = BigDecimal("10000.00"),
            startDate = LocalDate.now(),
            targetDate = LocalDate.now().minusDays(5)
        )

        whenever(userService.getAuthenticatedUserEntity()).thenReturn(testUser)

        assertThrows(BadRequestException::class.java) { savingsGoalService.createGoal(request) }
        verify(savingsGoalRepository, never()).save(any<SavingsGoal>())
    }

    @Test
    fun getGoalProgress_SuccessCalculations() {
        val startDate = LocalDate.now().minusMonths(3)
        val goal = SavingsGoal(
            id = 1L,
            goalName = "Car fund",
            targetAmount = BigDecimal("10000.00"),
            startDate = startDate,
            targetDate = LocalDate.now().plusYears(1),
            user = testUser
        )

        val expectedProgress = GoalProgressResponse(
            id = 1L,
            goalName = "Car fund",
            targetAmount = BigDecimal("10000.00"),
            currentProgress = BigDecimal("3000.00"),
            progressPercentage = 30.0,
            remainingAmount = BigDecimal("7000.00")
        )

        whenever(userService.getAuthenticatedUserEntity()).thenReturn(testUser)
        whenever(savingsGoalRepository.findByIdAndUserId(1L, 100L)).thenReturn(Optional.of(goal))

        whenever(transactionRepository.sumAmountByUserIdAndCategoryTypeSinceDate(100L, CategoryType.INCOME, startDate))
            .thenReturn(BigDecimal("5000.00"))
        whenever(transactionRepository.sumAmountByUserIdAndCategoryTypeSinceDate(100L, CategoryType.EXPENSE, startDate))
            .thenReturn(BigDecimal("2000.00"))

        whenever(savingsGoalMapper.toProgressResponse(eq(goal), any(), any(), any()))
            .thenReturn(expectedProgress)

        val progress = savingsGoalService.getGoalProgress(1L)

        assertNotNull(progress)
        assertEquals(BigDecimal("3000.00"), progress.currentProgress)
        assertEquals(30.0, progress.progressPercentage)
        assertEquals(BigDecimal("7000.00"), progress.remainingAmount)
    }
}
